libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1564,18 +1564,62 @@ bool ExpressionCompiler::visit(IndexAccess const& _indexAccess)
 				setLValueToStorageItem(_indexAccess);
 			break;
 		case DataLocation::Memory:
+			ArrayUtils(m_context).accessIndex(arrayType);
 			setLValue<MemoryItem>(_indexAccess, *_indexAccess.annotation().type, !arrayType.isByteArray());
 			break;
 		case DataLocation::CallData:
-			//@todo if we implement this, the value in calldata has to be added to the base offset
-			solUnimplementedAssert(!arrayType.baseType()->isDynamicallySized(), "Nested arrays not yet implemented.");
-			if (arrayType.baseType()->isValueType())
-				CompilerUtils(m_context).loadFromMemoryDynamic(
-					*arrayType.baseType(),
-					true,
-					!arrayType.isByteArray(),
-					false
+			if (arrayType.baseType()->isDynamicallyEncoded())
+			{
+				// stack layout: <base_ref> <length> <index>
+				ArrayUtils(m_context).accessIndex(arrayType, true, true);
+				// stack layout: <base_ref> <ptr_to_tail>
+
+				// returns the absolute offset of the accessed element in "base_ref"
+				m_context.appendInlineAssembly(
+					Whiskers(R"({
+					let rel_offset_of_tail := calldataload(ptr_to_tail)
+					if iszero(slt(rel_offset_of_tail, sub(sub(calldatasize(), base_ref), 0x1f))) { revert(0, 0) }
+					base_ref := add(base_ref, rel_offset_of_tail)
+				})").render(),
+					{"base_ref", "ptr_to_tail"}
 				);
+				// stack layout: <absolute_offset_of_tail> <garbage>
+
+				if (arrayType.baseType()->isDynamicallySized())
+				{
+					unsigned int baseEncodedSize = arrayType.baseType()->category() == Type::Category::Array ?
+						dynamic_cast<ArrayType const&>(*arrayType.baseType()).calldataStride() :
+						arrayType.baseType()->calldataEncodedSize();
+					solAssert(baseEncodedSize > 0, "");
+
+					// returns the absolute offset of the accessed element in "base_ref"
+					// and the length of the accessed element in "ptr_to_length"
+					m_context.appendInlineAssembly(
+						Whiskers(R"({
+							length := calldataload(base_ref)
+							base_ref := add(base_ref, 0x20)
+							if gt(length, 0xffffffffffffffff) { revert(0, 0) }
+							if gt(add(base_ref, mul(length, <baseEncodedSize>)), calldatasize()) { revert(0, 0) }
+						})")("baseEncodedSize", toCompactHexWithPrefix(baseEncodedSize)).render(),
+						{"base_ref", "length"}
+					);
+					// stack layout: <absolute_offset_of_element> <length_of_element>
+				}
+				else
+					m_context << Instruction::POP;
```

Comment:
Can you assert that we have a struct here?

Reply:
Actually I think what is missing here is that we still have enough room in the calldata to encode the size of the type (especially if it is a struct), right?

Reply:
Yeah, I was in fact wondering about this this morning :-). It may be that only statically sized arrays or structs end up here and those are encoded in-place and the bounds checks are in ``ArrayUtils::accessIndex``, but I'll double-check.

Reply:
The base type is dynamically encoded here, so on the stack is the tail pointer - that the tail pointer itself fits into calldata is verified above. If we want to check that the tail pointer does point to a region in calldata that is large enough for the type, we need an additional calldataload here just for the check - I'm wondering whether we need this check here or whether it's sufficient to check on access (which should happen, but only in the very end when actually accessing a value...).

Reply:
A bit confused here :-) - but you're right, we can and should add such a check here.

Reply:
Now I check against the calldatasize of the base type above already - that should be the correct check in all cases, right?